cpg-core/src/main/java/de/fraunhofer/aisec/cpg/helpers/SubgraphWalker.java

```diff
@@ -177,9 +177,7 @@ private static void flattenASTInternal(@NonNull Set<Node> list, @NonNull Node n)
     list.add(n);
 
     for (Node child : SubgraphWalker.getAstChildren(n)) {
-      if (!list.contains(child)) {
-        flattenASTInternal(list, child);
-      }
+      flattenASTInternal(list, child);
```

Comment:
The `list` has been used to detect loops in the AST which could originate from errors during translation or similar. With those changes, the argument `list` is no longer used for this purpose and I'm wondering if this could be an issue. Instead of the contains, we could also return if the `add` operation returns `false`.

Also, `list` is not a good name for the argument because it's actually a set, probably `result` or similar makes more sense in the current usage.

Reply:
The check `if (!list.contains(child))` was meant to avoid endless recursion. I don't see how this is ensured now. Of course, the set only contains each element once but we may try to add it multiple times, or not? I'd suggest to fix this by changing line 177 from `list.add(n)` to `if(!list.add(n)) return`. Or maybe I got something wrong?

Reply:
Ah yes, you are right. I was only adding the loop detection to the AST visitor (see VisitorTest), but not to the flattenAST.